lib/Maskings/Maskings.cpp

```diff
@@ -0,0 +1,309 @@
+                              std::vector<std::string>& path,
+                              VPackSlice const& data) {
+  for (auto const& entry : VPackArrayIterator(data)) {
+    if (entry.isObject()) {
+      VPackObjectBuilder ob(&builder);
+      addMaskedObject(collection, builder, path, entry);
+    } else if (entry.isArray()) {
+      VPackArrayBuilder ap(&builder);
+      addMaskedArray(collection, builder, path, entry);
+    } else {
+      std::string buffer;
+      builder.add(maskedItem(collection, path, buffer, entry));
+    }
+  }
+}
+
+void Maskings::addMaskedObject(Collection& collection, VPackBuilder& builder,
+                               std::vector<std::string>& path,
+                               VPackSlice const& data) {
+  for (auto const& entry : VPackObjectIterator(data, false)) {
+    std::string key = entry.key.copyString();
+    VPackSlice const& value = entry.value;
+
+    path.push_back(key);
+
+    if (value.isObject()) {
+      VPackObjectBuilder ob(&builder, key);
+      addMaskedObject(collection, builder, path, value);
+    } else if (value.isArray()) {
+      VPackArrayBuilder ap(&builder, key);
+      addMaskedArray(collection, builder, path, value);
+    } else {
+      std::string buffer;
+      builder.add(key, maskedItem(collection, path, buffer, value));
+    }
+
+    path.pop_back();
+  }
+}
+
+void Maskings::addMasked(Collection& collection, VPackBuilder& builder,
+                         VPackSlice const& data) {
+  if (!data.isObject()) {
+    return;
+  }
+
+  std::vector<std::string> path;
+  std::string dataStr("data");
+  VPackObjectBuilder ob(&builder, dataStr);
+
+  addMaskedObject(collection, builder, path, data);
+}
+
+void Maskings::addMasked(Collection& collection, basics::StringBuffer& data,
+                         VPackSlice const& slice) {
+  if (!slice.isObject()) {
+    return;
+  }
+
+  velocypack::StringRef dataStrRef("data");
+
+  VPackBuilder builder;
+
+  {
+    VPackObjectBuilder ob(&builder);
+
+    for (auto const& entry : VPackObjectIterator(slice, false)) {
+      velocypack::StringRef key = entry.key.stringRef();
+
+      if (key.equals(dataStrRef)) {
+        addMasked(collection, builder, entry.value);
+      } else {
+        builder.add(key, entry.value);
+      }
+    }
+  }
+
+  std::string masked = builder.toJson();
+  data.appendText(masked);
+  data.appendText("\n");
+}
+
+void Maskings::mask(std::string const& name, basics::StringBuffer const& data,
+                    basics::StringBuffer& result) {
+  result.clear();
+
+  auto const itr = _collections.find(name);
+
+  if (itr == _collections.end()) {
+    result.copy(data);
+    return;
+  }
+
+  result.reserve(data.length());
+
+  char const* p = data.c_str();
+  char const* e = p + data.length();
+  char const* q = p;
+
+  while (p < e) {
+    while (p < e && (*p != '\n' && *p != '\r')) {
+      ++p;
+    }
+
+    std::shared_ptr<VPackBuilder> builder = VPackParser::fromJson(q, p - q);
```

Comment:
This will create a lot of Builder objects. I think it is ok to create new Parser objects inside the loop, but all parsers can share the same Builder object, which can be declared outside the loop and recycled inside the loop for each document.